compiler/next/include/chpl/AST/BaseAST.h

```diff
@@ -0,0 +1,215 @@
+/*
+ * Copyright 2021 Hewlett Packard Enterprise Development LP
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef CHPL_AST_BASEAST_H
+#define CHPL_AST_BASEAST_H
+
+#include "chpl/AST/ASTTag.h"
+#include "chpl/AST/ID.h"
+#include "chpl/AST/ASTTypes.h"
+#include "chpl/Util/memory.h"
+
+#include <cassert>
+
+namespace chpl {
+namespace ast {
+
+/**
+  This is the base class for AST types.
+
+  Every AST class has:
+   * a tag (indicating which AST class it is)
+   * an ID (a sort of scoped location used as a key in maps)
+   * a list of child Exprs
+
+  The list of child Exprs is included in BaseAST to allow for
+  generic tree traversals of the AST.
+
+  someAst->isCallExpr() / someAst->toCallExpr() are available
+  and generated for all AST types.
+
+  std::less is defined for every AST class and it compares IDs.
+
+ */
+class BaseAST {
+ friend class Builder;
+
+ private:
+  asttags::ASTTag tag_;
```

Comment:
I'd prefer to refer to `ASTTag` without the namespace prefix.

Reply:
Sure. I'll do that in a follow-up.